Issue13 count motifs

stinea98 opened the pull request:
### Description
This function calculates all the occuring motifs in a set of sequences in a specified length and returns them as a dict.

Fixes # 13

### Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality
      to not work as expected)
- [ ] Documentation update

### Checklist

Please carefully read these items and tick them off if the statements are true
_or do not apply_.

- [x] I have performed a self-review of my own code
- [x] My code follows the existing coding style, lints and generates no new
      warnings
- [x] I have added type annotations to all function/method signatures, and I
      have added type annotations for any local variables that are non-trivial,
      potentially ambiguous or might otherwise benefit from explicit typing.
- [x] I have commented my code in hard-to-understand areas
- [x] I have added ["Google-style docstrings"] to all new modules, classes,
      methods/functions or updated previously existing ones
- [x] I have added tests that prove my fix is effective or that my feature
      works
- [x] New and existing unit tests pass locally with my changes and I have not
      reduced the code coverage relative to the previous state
- [x] I have updated any sections of the app's documentation that are affected
      by the proposed changes

Commit:
add test.md

Commit:
added ideas for function description to test.md

lukasheuberger committed:
issue13 outline + function

added more ideas to solve problem and added function docstring

Commit:
updated issue13.md

lukasheuberger committed:
create issue13.py

Commit:
updated md file and code

lukasheuberger committed:
Update issue13.py

Commit:
minor changes to code

lukasheuberger committed:
added unittest, renamed module, more documentation

Commit:
compatible with all 3 linters

lukasheuberger committed:
unittest changes, docstring minor changes

- added mock package to unittest
- adapted docstring with datatypes
- added pull-request tempalte

Commit:
deleted .md file

uniqueg committed:
Merge branch 'dev' into issue13_count-motifs

stinea98 commented:
this is from Lukas Heuberger and Stine Anzböck.

lukasheuberger committed:
minor changes for pull request

- typing for variables
- added module doctsring and minor adjustment in function doctsring
- linting of test_functions

uniqueg commented on .idea/.gitignore:
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
```
Please add `.idea` to the `.gitignore` file in the project root as we don't want to version-control your personal editor files. You will also need to explicitly remove the files that were already added (all files in `.idea` folder). Find more info online, e.g., here: https://docs.github.com/en/free-pro-team@latest/github/using-git/ignoring-files

This comment applies to the next couple of files, too.

uniqueg commented on htsinfer/infer_count_motifs.py:
```diff
@@ -0,0 +1,91 @@
+from typing import Dict
+
+
+def count_motifs(input_sequences: list, min_motif_length: int,
```
Using the builtin `list` and `dict` types for type hinting is only really supported since Python 3.9 (which came out in October). For backwards compatibility it is better to use `List` and `Dict` which you need to explicitly import from the `typing` module.

uniqueg commented on htsinfer/infer_count_motifs.py:
```diff
@@ -0,0 +1,91 @@
+    the specified length and their occurrence.
+
+        Args:
+            input_sequences (list): list of sequences
```
Since you are using type hints, you can (and probably should) omit the type info from the docstrings, as documentation tooling can automatically parse the type hints and add them to API docs. The problem about adding info to comments rather than in the code is that over time, comments and code tend to diverge, because people forget to update the comments. It's always better to have a single source of truth, and the code is the only really source of truth...

uniqueg commented on htsinfer/infer_count_motifs.py:
```diff
@@ -0,0 +1,91 @@
+            sorted_dict (dict): paired motif frequency {motif: frequency }
+
+        Raises:
+            TypeError: input_sequences is not a list
```
If you put named entities like input parameters in docstrings, it is good practice to put them between backticks so that they are rendered and possibly even linked by autodocumentation tooling. So better to write

```text
TypeError: `input_sequences` is not a list
```

etc.

uniqueg commented on htsinfer/infer_count_motifs.py:
```diff
@@ -0,0 +1,91 @@
+
+    for sequence_no in range(0, len(input_sequences)):
+        seq = input_sequences[sequence_no]
+        user_choice = "0"
```
If it can be at all avoided, it's not the best idea to make your script interactive because you cannot use it in pipelines anymore and even if run standalone, it's a bit annoying to start a script, then the script runs, possibly for x minutes, then the script stops and waits for user input. It's just not the best user experience. It is better to add a parameter `non-nucleotide characters` to the function and accept a string value that could be either `ignore_chars`, `ignore_seqs` and `include`, then set a default for that parameter (probably `ignore_seqs`).

uniqueg commented on htsinfer/infer_count_motifs.py:
```diff
@@ -0,0 +1,91 @@
+            input_sequences (list): list of sequences
+            min_motif_length (int): minimal length of motif
+            max_motif_length (int): maximal length of motif
+            nucleic_acid (string): type of nucleic acid (dna, rna)
```
Indicate with single or double quotes that 'dna' and 'rna' are actually the expected string arguments for this parameter

uniqueg replied:
Useful to set a default, preferentially `dna`. Always try to make your functions easy to use

uniqueg commented on htsinfer/infer_count_motifs.py:
```diff
@@ -0,0 +1,91 @@
+            ValueError: max_motif_length is longer than the shortest list
+    """
+
+    if not isinstance(input_sequences, list):
```
The input you'll get in this function will be passed from another function, and the type hints will make sure that we will realize if what we pass is not of the correct type. Thus, when using type hints, it is really only necessary to validate actual raw user input.

Also note that you might feel that, having done your type checks, you are going to be safe when actually using the variables, but in Python, the type of a variable can change between one value and another. Even if you're sure that you don't put any code in between your validation and the place(s) where you use a variable that may change the type, someone else _may_ do so in the future, and the validation might give a false sense of security. If you don't want to have your script just fail in a case like that, the best thing is to put your potentially failing code in a `try - except` block and specifically catch any exceptions that you can think of.

uniqueg commented on htsinfer/infer_count_motifs.py:
```diff
@@ -0,0 +1,91 @@
+
+def count_motifs(input_sequences: list, min_motif_length: int,
+                 max_motif_length: int, nucleic_acid: str) -> dict:
+    """Function that calculates the occurrence of all motifs in one or
```
Best to keep these short, precise and free of redundancy. The function signature and context make it clear that this is a function. And the `Returns` section specifies in more detail what is returned.

Better would be something like: "Count motifs on one or more nucleotide sequences"

uniqueg commented on htsinfer/infer_count_motifs.py:
```diff
@@ -0,0 +1,91 @@
+        raise ValueError('nucleic_acid is not in specified range (dna/rna)!')
+
+    if max_motif_length > len(min(input_sequences)):
+        raise ValueError('max_motif_length is longer than the shortest list')
```
Didn't you want to write:
> is longer than the shortest ~list~ _sequence_
?

Still, I don't really get why this is a problem. Shouldn't you make sure the _minimal_ motif length is not bigger than the shortest sequence? I think it's fine if the maximum motif length is never found

uniqueg commented on htsinfer/infer_count_motifs.py:
```diff
@@ -0,0 +1,91 @@
+    if not input_sequences:
+        raise TypeError('input_sequences is not a list of sequences!')
+
+    if not isinstance(min_motif_length, int):
```
Rather than checking the type I'd rather check if it is a _positive_ integer and that `max` isn't smaller than `min`

uniqueg commented on htsinfer/infer_count_motifs.py:
```diff
@@ -0,0 +1,91 @@
+                else:
+                    print('motif not valid: ', motif)
+
+    sorted_dict = dict(
```
I wouldn't return a sorted dict, because it takes time to sort and sorting isn't always desired. Therefore it's probably better to let the _consumer_ of the dictionary sort it, because they know better if sorting is needed or not.

uniqueg commented on tests/test_count_motifs.py:
```diff
@@ -0,0 +1,104 @@
+"""Unit tests for infer_single_paired.py"""
```
The tests are really nice! Well done!  :)

(though probably need some changes after addressing the comments above, sorry about that)

uniqueg commented:
@stinea98 & @lukasheuberger: sorry, forgot to review your PR earlier. It's really nice and thorough work! Still added quite a few comments for improvement. Would be great if you could address these

Commit:
resolved some of the conversations & linter-approved again

lukasheuberger committed:
removed user-input and improved error handling

- updated gitignore with .idea

lukasheuberger commented:
@uniqueg thank you for your feedback. we tried to address these points and improved our code. however, it still fails to auto-merge because of the error "R0912: Too many branches (20/12) (too-many-branches)". I assume it is because we have too many if-statements for the error-handling. should we try to get rid of more of these or is it still ok if we get this error?

Commit:
minor fix in function description

uniqueg commented:
@lukasheuberger Yes, please have a try. If you don't manage, you can check online on how to disable specific Pylint warnings, either per line or by adding the error code to the [Pylint config file](https://github.com/zavolanlab/htsinfer/blob/dev/pylint.cfg). Goes for the other issue as well.

lukasheuberger committed:
reduced if-else and added tests

- added append_motif function to reduce if-else
- added tests for improved code coverage

uniqueg commented on htsinfer/infer_count_motifs.py:
```diff
@@ -0,0 +1,98 @@
+        # and continues with the next sequence
+        if not all(base in valid_bases for base in seq) \
+                and non_nucleotide_characters == 'ignore_seqs':
+            print('sequence', sequence_no+1, 'ignored')
```
Use `print()` only to debug, use `logging` library for log messages

mzavolan commented on htsinfer/infer_count_motifs.py:
```diff
@@ -0,0 +1,98 @@
+
+                if all(char in valid_bases for char in motif) or \
+                        non_nucleotide_characters == 'include':
+                    motif_freq = append_motif(motif, motif_freq)
```
The dictionary of motif counts should not be rewritten again and again. Only the counts of motifs should be updated. One can use defaultdict to take care of initialization.

uniqueg commented:
Looks pretty good. Resolve conflicts, review and resolve open comments, make sure all tests are reasonable and merge.

uniqueg commented:
Using PR #55 instead
